core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -57,16 +64,29 @@ private[kafka] abstract class SubSourceLogic[K, V, Msg](
     }
   }
 
-  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
+  private def pumpCB = getAsyncCallback[Set[TopicPartition]] { tps =>
     pendingPartitions ++= tps.filter(!partitionsInStartup.contains(_))
     pump()
   }
 
-  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
-    pendingPartitions --= tps
-    partitionsInStartup --= tps
-    tps.flatMap(subSources.get).foreach(_.shutdown())
-    subSources --= tps
+  private implicit val askTimeout = Timeout(5000, TimeUnit.MILLISECONDS)
+  def partitionAssignedCB(tps: Set[TopicPartition]) = {
+    implicit val ec = materializer.executionContext
+    getOffsetsOnAssign.fold(pumpCB.invoke(tps)) { getOffsets =>
+      getOffsets(tps).flatMap { offsets =>
+        consumer.ask(KafkaConsumerActor.Internal.Seek(offsets)).map(_ => pumpCB.invoke(tps))
```

Comment:
what should happen if ask timeout? ok to ignore?

Reply:
I think we'll have to fail the stream if the ask times out. We can't proceed correctly if we don't know if the `seek` happened. We could maybe retry once before failing. Sending this message won't result in any remote network calls, it's just a local message send to the `KafkaConsumerActor` and a call to `KafkaConsumer.seek` which doesn't trigger a call to the broker, so it shouldn't really ever time out anyway. If it does time out, something probably went wrong with the `KafkaConsumerActor`.

I'll add some logic to stop the stream if it happens.

Reply:
KafkaConsumerActor is sometimes blocking on the `poll`. Perhaps increase the timeout further and if it happens fail the stream. No need for retry.